Multicall docs (#1227)
software-mansion/protostar: website/docs/tutorials/08-interacting-with-starknet/07-multicall.md

# Multicall

## Overview

> Multicall allows executing multiple calls while providing [atomicity](https://en.wikipedia.org/wiki/Atomicity_(database_systems)). If one call fails, the entire operation is aborted. Use it to prevent leaving your system in an inconsistent state. 

Reviewer: ```suggestion
Multicall executes multiple calls as a single transaction. If one call fails, the entire operation is aborted. Use it to prevent leaving your system in an inconsistent state. 
```
Atomicity does not relate to consistency. Atomicity relates to not beaing interleaved by another transaction. https://en.wikipedia.org/wiki/ACID

Author: Consistency is a property of an application that is achieved using Atomicity and Isolation, hence Atomicity is related to Consistency.

Atomicity is more about abortability.

## Usage example

Before you run [`protostar multicall`](/docs/cli-reference#multicall), you need to [create an account](./05-deploy-account.md) in order to [sign the transaction](./06-signing.md). If you want to deploy contract within a multicall, you need to [declare](./03-declare.md) it first.

> Then, create a toml file containing call [array of tables](https://toml.io/en/v1.0.0#array-of-tables).

Reviewer: ```suggestion
Then, create a toml file containing the multicall [array of tables](https://toml.io/en/v1.0.0#array-of-tables).
```

Author: "multicall array of tables" doesn't make sense. It's `[[call]]` not `[[multicall]]`.

Reviewer: So it should be plural i suppose?

`containing [array_of_tables] including a sequence of calls

Author: `containing calls defined with [array of tables]`

```toml title="calls.toml"
[[call]]
type = "deploy"
class-hash = 0xDEADBEEF
inputs = []
id = "my_contract"

[[call]]
type = "invoke"
contract-address = "$my_contract"
function = "increase_balance"
inputs = [42]
```

> Protostar supports two types of calls — `deploy` and `invoke`, which accept similar arguments to [`deploy command`](/docs/cli-reference#deploy) and [`invoke command`](/docs/cli-reference#invoke).

Reviewer: ```suggestion
Protostar supports two types of calls — `deploy` and `invoke`, which take similar arguments to [`deploy command`](/docs/cli-reference#deploy) and [`invoke command`](/docs/cli-reference#invoke).
```

> You can use contract address from a previous call by using references (`$DEPLOY_CALL_ID`). References can be used in `contract-address` and `inputs` attributes as demonstrated in the example above.

Reviewer: If it relates to example above can it move it above it? We have one example below as well and this is misleading

Author: Moved

# Multicall

## Overview
Multicall executes multiple calls as a single transaction. If one call fails, the entire operation is aborted. Use it to prevent leaving your system in an inconsistent state. 

## Usage example

Before you run [`protostar multicall`](/docs/cli-reference#multicall), you need to [create an account](./05-deploy-account.md) in order to [sign the transaction](./06-signing.md). If you want to deploy contract within a multicall, you need to [declare](./03-declare.md) it first.

> Then, create a toml file containing calls defined as[array of tables](https://toml.io/en/v1.0.0#array-of-tables).

Reviewer: ```suggestion
Then, create a toml file containing calls defined as [array of tables](https://toml.io/en/v1.0.0#array-of-tables).
```
